### DISCUSSION TITLE
Sixel / Inline Image Protocol Support

### DISCUSSION THREAD

**Discussion by ShaunLawrie** [spectreconsole/spectre.console#1671] [Ideas]
@trackd and I've been playing around with Sixel seeing as it's finally going to be supported in the next Windows Terminal release. I think it could be cool to extend the functionality in the Spectre.Console.ImageSharp package to handle outputting sixel data for terminals that support it. I've used spectre console as my slide deck engine for quite a few presentations and being able to rick roll an audience in higher definition is something I've craved...

I have a rough working copy of sixel in my fork at the moment at https://github.com/spectreconsole/spectre.console/compare/main...ShaunLawrie:spectre.console:main

The image class is SixelImage and is used the same as CanvasImage e.g.

```csharp
var image = new SixelImage("pikachu.png");
image.MaxWidth = 40 // this is cell width not pixels
```

It seems to work with alignment/embedding so it's in a decent PoC state.
![image](https://github.com/user-attachments/assets/bed71a03-485d-4797-89db-5c46b885bea6)

This implementation dumps the raw sixel data as a controlcode segment so it has no dimensions in Spectre.Console. To allow the image to have dimensions, the controlcode segment is directly followed by a blank canvas of the same dimensions. This may feel a bit odd but it was the way I could see adding it without making loads of changes in spectre.console core.

## Things to consider:
- Is it worth upstreaming sixel into imagesharp instead of doing it here? https://github.com/SixLabors/ImageSharp/discussions/2796
- What other image protocols would need to be supported to stop people raising bugs like "it doesn't work in wezterm", @trackd has been looking at this stuff in https://github.com/trackd/Sixel. Afaik Sixel is the most complicated to implement which has been done in this poc.
- Would an approach of "render in high def and fall back to canvasimage" be a good option?
- Does anyone else actually want this?
- Sixel data doesn't align to character cells in any way it depends on the cell/pixel ratio of the console host that you can interrogate via control codes. Is it best to always align images to cells to avoid having to deal with partial cells leaving renderables looking misaligned?
- Dumping sixel data to the console is significantly faster than dumping a whole page of ansi cells with coloring, I was surprised by this and it makes things like animations possible but I'm not sure how that would fit in spectre console.

**Comment by patriksvensson** [comment]
@ShaunLawrie Just saw this. Will look at it closer tomorrow.

**Reply by powercode** [reply]
Eagerly awaiting this!!

Do you have any progress to share?
